File: backend/app/streaming/sse.py
"""SSE event taxonomy — the contract between agent and frontend.

Each event has a `type` and a `data` payload. Wire format: `data: {json}\n\n`.
"""
from __future__ import annotations

import json
from enum import Enum
from typing import Any
from pydantic import BaseModel, Field


class EventType(str, Enum):
    THINKING_DELTA = "thinking_delta"
    AGENT_STEP_START = "agent_step_start"
    AGENT_STEP_END = "agent_step_end"
    TOOL_CALL = "tool_call"
    MODEL_PROPOSAL = "model_proposal"
    JURY_VERDICT = "jury_verdict"
    TIEBREAKER_RESOLVED = "tiebreaker_resolved"
    VALIDATOR_CHECK = "validator_check"
    RANKED_CARD = "ranked_card"
    CITATION = "citation"
    TEXT_DELTA = "text_delta"
    EXCLUSION = "exclusion"
    STREAM_COMPLETE = "stream_complete"
    ERROR = "error"


class StreamEvent(BaseModel):
    type: EventType
    data: dict[str, Any] = Field(default_factory=dict)

    def to_sse(self) -> str:
        payload = json.dumps({"type": self.type.value, "data": self.data})
        return f"data: {payload}\n\n"


def thinking(step_id: str, text: str) -> StreamEvent:
    return StreamEvent(type=EventType.THINKING_DELTA, data={"step_id": step_id, "text": text})


def agent_step_start(name: str, model: str | None = None, label: str = "") -> StreamEvent:
    data: dict[str, Any] = {"name": name, "label": label}
    if model:
        data["model"] = model
    return StreamEvent(type=EventType.AGENT_STEP_START, data=data)


def agent_step_end(name: str, latency_ms: int = 0, label: str = "", meta: dict | None = None) -> StreamEvent:
    data: dict[str, Any] = {"name": name, "latency_ms": latency_ms, "label": label}
    if meta:
        data["meta"] = meta
    return StreamEvent(type=EventType.AGENT_STEP_END, data=data)


def tool_call(name: str, input: Any, output_summary: str = "", runtime_ms: int = 0,
              meta: dict | None = None) -> StreamEvent:
    data: dict[str, Any] = {
        "name": name, "input": input, "output_summary": output_summary, "runtime_ms": runtime_ms,
    }
    if meta:
        data["meta"] = meta
    return StreamEvent(type=EventType.TOOL_CALL, data=data)


def model_proposal(slot: str, model: str, vendor: str, content: str) -> StreamEvent:
    return StreamEvent(type=EventType.MODEL_PROPOSAL, data={
        "slot": slot, "model": model, "vendor": vendor, "content": content,
    })


def jury_verdict(claim_id: str, claim_text: str, judges: list[dict],
                 agreement: dict, final_verdict: str) -> StreamEvent:
    return StreamEvent(type=EventType.JURY_VERDICT, data={
        "claim_id": claim_id, "claim_text": claim_text,
        "judges": judges, "agreement": agreement, "final_verdict": final_verdict,
    })


def tiebreaker_resolved(claim_id: str, model: str, rationale: str, final_verdict: str) -> StreamEvent:
    return StreamEvent(type=EventType.TIEBREAKER_RESOLVED, data={
        "claim_id": claim_id, "model": model, "rationale": rationale, "final_verdict": final_verdict,
    })


def validator_check(model: str, status: str, message: str,
                    broken_offsets: list | None = None) -> StreamEvent:
    data: dict[str, Any] = {"model": model, "status": status, "message": message}
    if broken_offsets:
        data["broken_offsets"] = broken_offsets
    return StreamEvent(type=EventType.VALIDATOR_CHECK, data=data)


def ranked_card(rank: int, facility_id: str, name: str, location: str,
                distance_km: float | None, type_: str, trust_score: float,
                prose: str, citation_ids: list[str], primary_claim_id: str,
                meta: dict | None = None) -> StreamEvent:
    data: dict[str, Any] = {
        "rank": rank, "facility_id": facility_id, "name": name, "location": location,
        "distance_km": distance_km, "type": type_, "trust_score": trust_score,
        "prose": prose, "citation_ids": citation_ids, "primary_claim_id": primary_claim_id,
    }
    if meta:
        data["meta"] = meta
    return StreamEvent(type=EventType.RANKED_CARD, data=data)


def citation(citation_id: str, facility_id: str, column: str, char_start: int,
             char_end: int, excerpt: str) -> StreamEvent:
    return StreamEvent(type=EventType.CITATION, data={
        "citation_id": citation_id, "facility_id": facility_id, "column": column,
        "char_start": char_start, "char_end": char_end, "excerpt": excerpt,
    })


def text(delta: str) -> StreamEvent:
    return StreamEvent(type=EventType.TEXT_DELTA, data={"text": delta})


def exclusion(facility_id: str, name: str, location: str, type_: str,
              reason: str, verdict: str) -> StreamEvent:
    return StreamEvent(type=EventType.EXCLUSION, data={
        "facility_id": facility_id, "name": name, "location": location, "type": type_,
        "reason": reason, "verdict": verdict,
    })


def stream_complete(recommendation_count: int, exclusion_count: int,
                    total_latency_ms: int) -> StreamEvent:
    return StreamEvent(type=EventType.STREAM_COMPLETE, data={
        "recommendation_count": recommendation_count,
        "exclusion_count": exclusion_count,
        "total_latency_ms": total_latency_ms,
    })


def error(message: str, stage: str | None = None) -> StreamEvent:
    data: dict[str, Any] = {"message": message}
    if stage:
        data["stage"] = stage
    return StreamEvent(type=EventType.ERROR, data=data)

File: frontend/lib/sse.ts
/**
 * Sanjeevani — SSE event contract.
 *
 * This module is the source of truth for the wire format between the FastAPI
 * backend and the Next.js frontend. The discriminated union below tells you
 * exactly what JSON shape each event must produce; the parser is the only
 * frontend code that reads raw SSE frames.
 *
 * Partner backend reference: docs/sse-event-contract.md (mirrors this file).
 */
import type { RecommendedFacility, ExcludedFacility, Verdict } from "@/lib/types";

export type StreamEvent =
  | { type: "thinking_delta"; data: { text: string } }
  | {
      type: "agent_step_start";
      data: { step_id: string; name: string; label: string };
    }
  | {
      type: "agent_step_end";
      data: {
        step_id: string;
        name: string;
        summary?: string;
        /** Mono-text body shown under the step label — typically a SQL
         *  snippet, a numeric breakdown, or whatever rich detail the node
         *  produced. Free-form, multi-line allowed. */
        detail?: string;
        duration_ms?: number;
      };
    }
  | {
      type: "tool_call";
      data: {
        tool: string;
        input?: Record<string, unknown>;
        output_summary?: string;
        duration_ms?: number;
      };
    }
  | {
      type: "model_proposal";
      data: { proposer_id: "A" | "B"; vendor: string; title: string; text: string };
    }
  | { type: "consensus_resolved"; data: ConsensusResolvedData }
  | { type: "validator_pass"; data: { title: string; body: string; passed: boolean } }
  | { type: "text_delta"; data: { text: string } }
  | {
      type: "citation";
      data: {
        citation_id: string;
        facility_id: string;
        column: string;
        char_start: number;
        char_end: number;
        excerpt: string;
      };
    }
  | {
      type: "recommendations_ready";
      data: {
        facilities: RecommendedFacility[];
        excluded: ExcludedFacility[];
        pipeline_ms: number;
        candidates_considered: number;
      };
    }
  | { type: "error"; data: { message: string } };

export interface ConsensusResolvedData {
  /** The claim being verdicted, e.g. "performs emergency appendectomy". */
  claim: string;
  /** Optional display title, e.g. "Emergency surgery — Sadar Hospital". */
  title?: string;
  verdict: Verdict;
  /** Number of judges agreeing with the consensus verdict (0–3). */
  agreement: 0 | 1 | 2 | 3;
  dissent: boolean;
  judges: Array<{
    /** Display name, e.g. "Claude Sonnet 4.6". */
    name: string;
    /** Vendor + provider, e.g. "Anthropic · Bedrock". */
    vendor: string;
    verdict: Verdict;
    /** 0–1 confidence score. */
    confidence: number;
    /** Judge's quoted evidence sentence. */
    excerpt: string;
    /** Set on dissenting judges only. */
    dissent_note?: string;
  }>;
  /** Present only when all three judges initially disagreed. */
  tiebreaker?: { model: string; verdict: Verdict; reasoning: string };
}

export type StreamEventType = StreamEvent["type"];

/**
 * Event types the existing backend (`backend/app/streaming/sse.py`) already
 * emits. Used by `useEventStream` to decide which event types it must
 * synthesize from baked demo data when the live stream omits them.
 */
const LIVE_ONLY_TYPES: ReadonlySet<StreamEventType> = new Set<StreamEventType>([
  "thinking_delta",
  "agent_step_start",
  "agent_step_end",
  "tool_call",
  "text_delta",
  "citation",
  "error",
]);

export function isLiveOnlyEvent(t: StreamEventType): boolean {
  return LIVE_ONLY_TYPES.has(t);
}

/** Parses a single SSE frame ("data: {...}") into a typed event, or null. */
export function parseSSEFrame(frame: string): StreamEvent | null {
  if (!frame.startsWith("data: ")) return null;
  const json = frame.slice(6);
  try {
    return JSON.parse(json) as StreamEvent;
  } catch {
    return null;
  }
}

/**
 * POST a query to the backend and yield SSE events as they arrive.
 * The hook layer (`useEventStream`) merges live + baked demo data.
 *
 * Pass an AbortSignal to cancel the in-flight fetch (used to discard a
 * superseded run when the user changes the active query).
 */
export async function* streamQuery(
  query: string,
  signal?: AbortSignal,
): AsyncGenerator<StreamEvent> {
  const apiUrl = process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000";
  const resp = await fetch(`${apiUrl}/query`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ query }),
    signal,
  });
  if (!resp.ok || !resp.body) {
    throw new Error(`HTTP ${resp.status}`);
  }
  const reader = resp.body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";

  try {
    while (true) {
      const { value, done } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });
      let idx;
      while ((idx = buffer.indexOf("\n\n")) !== -1) {
        const frame = buffer.slice(0, idx);
        buffer = buffer.slice(idx + 2);
        const ev = parseSSEFrame(frame);
        if (ev) yield ev;
      }
    }
  } finally {
    reader.cancel().catch(() => {});
  }
}
